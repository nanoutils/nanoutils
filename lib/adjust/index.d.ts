import {
  CurriedFunction1,
  CurriedFunction2,
  Index,
  Fn,
  MixedArray
} from '../../misc/types/fn'

declare function adjust<T1, T2>(fn: Fn<T1, T2>): CurriedFunction2<number, Array<T1>, MixedArray<T1, T2>>
declare function adjust<T1, T2>(fn: Fn<T1, T2>, index: Index): CurriedFunction1<Array<T1>, MixedArray<T1, T2>>
declare function adjust<T1, T2>(fn: Fn<T1, T2>, index: Index, array: Array<T1>): MixedArray<T1, T2>

export default adjust