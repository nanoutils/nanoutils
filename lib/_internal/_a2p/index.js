export default (a) =>
  Array.isArray(a)
    ? a.join('.')
    : typeof a === 'string' || typeof a === 'number'
    ? '' + a
    : ''
