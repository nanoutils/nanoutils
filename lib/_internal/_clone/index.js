export function shallowCloneObjectAndArrays(obj) {
  var res = {}
  if (typeof obj !== 'object') {
    return res
  }
  if (Array.isArray(obj)) {
    return obj.slice()
  }
  for (var k in obj) {
    res[k] = obj[k]
  }
  return res
}
