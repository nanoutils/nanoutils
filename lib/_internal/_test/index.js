/**
 * Check for accepting number of arguments for given function
 * @param {function} f - function to check
 * @param {number} n - number of accepting arguments
 * @param {array} args - example of arguments to check
 */
export function expectNumberOfArgs(f, n, args) {
  let k = n
  while (k > 0) {
    expect(typeof f).toBe('function')
    f = f(args[n - k--])
  }
  expect(typeof f).not.toBe('function')
}

export function expectExactNumberOfArgs(f, args) {
  expect(typeof f.apply(f, args)).not.toBe('function')
}
