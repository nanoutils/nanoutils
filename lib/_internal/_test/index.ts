import { expect } from "earljs";

/**
 * Check for accepting number of arguments for given function
 * @param {function} f - function to check
 * @param {number} n - number of accepting arguments
 * @param {array} args - example of arguments to check
 */
export function expectNumberOfArgs<Args extends unknown>(
  f: (args: Args) => unknown,
  n: number,
  args: Args[]
) {
  let k = n;
  let g: (args: Args) => unknown = f;
  while (k > 0) {
    expect(g).toBeA(Function);
    g = g(args[n - k--]) as (args: Args) => unknown;
  }
  expect(g).not.toBeA(Function);
}

export function expectExactNumberOfArgs<Args extends unknown>(
  f: (args: Args) => unknown,
  args: Args[]
) {
  expect(f.apply(f, args)).not.toBeA(Function);
}
