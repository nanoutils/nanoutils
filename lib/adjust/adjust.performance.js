const { generators: { getSameLengthArrays }, time: { TIME_1M }, types: { TYPE_SHE } } = require('../_internal/helpers/performance')

module.exports = function() {
  const [start, half, end] = [0, TIME_1M / 2, TIME_1M - 1]
  const [arr1, arr2, arr3] = getSameLengthArrays()
  const add10 = x => x + 10

  return {
    type: TYPE_SHE,
    argss: [
      [add10, start, arr1], // first element
      [add10, half, arr2], // middle element
      [add10, end, arr3] // last element
    ]
  }
}
