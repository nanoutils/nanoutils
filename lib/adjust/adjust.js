export default function adjust(fn, index, array) {
  if (process.env.NODE_ENV !== 'production') {
    require('../_internal/_error').checkTypes('adjust', {
      fn: [fn, ['function'], typeof fn !== 'function'],
      index: [index, ['number'], typeof index !== 'number'],
      arr: [array, ['array'], !Array.isArray(array)]
    })
  }
  var result = array.slice()
  if (!(index in array)) return result
  result[index] = fn(result[index])
  return result
}
