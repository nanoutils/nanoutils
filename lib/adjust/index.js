import curry3 from '../_internal/_curry3'
import adjust from './adjust'

export default curry3(adjust)
