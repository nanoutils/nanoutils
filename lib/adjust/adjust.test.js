import { expectNumberOfArgs } from '../_internal/_test'
import adjust from '.'
import add from '../add'
import multiply from '../multiply'

const addTen = add(10)
const multiplyByTwo = multiply(2)

test('it accepts exact 3 arguments', () => {
  expectNumberOfArgs(
    adjust,
    3,
    [addTen, 1, [0, 1, 2]]
  )
})

test('applies function to array item by index', () => {
  expect(adjust(addTen, 2, [2, 3, 4, 5, 6])).toEqual([2, 3, 14, 5, 6])
  expect(adjust(multiplyByTwo, 4, [2, 3, 4, 5, 6])).toEqual([2, 3, 4, 5, 12])
})

test('return array itself if no element by index does not exist', () => {
  expect(adjust(addTen, 10, [2, 3, 4, 5, 6])).toEqual([2, 3, 4, 5, 6])
  expect(adjust(multiplyByTwo, 10, [2, 3, 4, 5, 6])).toEqual([2, 3, 4, 5, 6])
})

test('returns copy even if index out of bounds', () => {
  const arr = [1, 2, 3]
  expect(adjust(addTen, 4, arr) === arr).toBeFalsy()
})
