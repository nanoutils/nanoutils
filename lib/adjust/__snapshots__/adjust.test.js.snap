// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks dev errors 1`] = `"[Nanoutils] adjust(fn, *, *) accepts a function, number given"`;

exports[`checks dev errors 2`] = `"[Nanoutils] adjust(*, index, *) accepts a number, boolean given"`;

exports[`checks dev errors 3`] = `"[Nanoutils] adjust(*, *, arr) accepts an array, number given"`;
